bundles/org.openhab.binding.icalendar/src/main/java/org/openhab/binding/icalendar/internal/handler/PullJob.java

```diff
@@ -0,0 +1,203 @@
+     * @param password Optional password for basic auth. Must be set together with an username.
+     * @param destination The destination the downloaded calendar should be saved to.
+     * @param listener The listener that should be fired when update succeed.
+     */
+    public PullJob(HttpClient httpClient, URI sourceURI, @Nullable String username, @Nullable String password,
+            File destination, @Nullable CalendarUpdateListener listener) {
+        this.httpClient = httpClient;
+        this.sourceURI = sourceURI;
+        if (username != null && password != null) {
+            authentication = new BasicAuthentication.BasicResult(this.sourceURI, username, password);
+        } else {
+            authentication = null;
+        }
+        this.destination = destination;
+        this.listener = listener;
+    }
+
+    @Override
+    public void run() {
+        Request request = httpClient.newRequest(sourceURI).followRedirects(true).method(HttpMethod.GET);
+        Authentication.@Nullable Result currentAuthentication = authentication;
+        if (currentAuthentication != null) {
+            currentAuthentication.apply(request);
+        }
+
+        InputStreamResponseListener asyncListener = new InputStreamResponseListener();
+        request.send(asyncListener);
+
+        Response response;
+        try {
+            response = asyncListener.get(HTTP_TIMEOUT_SECS, TimeUnit.SECONDS);
+        } catch (InterruptedException | TimeoutException | ExecutionException e1) {
+            logger.warn("Response for calendar request could not be retrieved.");
+            return;
+        }
+
+        if (response.getStatus() != HttpStatus.OK_200) {
+            logger.warn("Response status for getting \"{}\" was {} instead of 200. Ignoring it.", sourceURI,
+                    response.getStatus());
+            return;
+        }
+
+        String responseLength = response.getHeaders().get(HttpHeader.CONTENT_LENGTH);
+        if (responseLength != null) {
+            try {
+                if (Integer.parseInt(responseLength) > HTTP_MAX_CALENDAR_SIZE) {
+                    logger.warn("Calendar is too big ({} bytes), aborting request", responseLength);
+                    response.abort(new ResponseTooBigException());
+                    return;
+                }
+            } catch (NumberFormatException e) {
+                logger.debug(
+                        "While requesting calendar Content-Length was set, but is malformed. Falling back to read-loop.",
+                        e);
+            }
+        }
+
+        File tmpTargetFile;
+        FileOutputStream tmpOutStream = null;
+        try {
+            tmpTargetFile = File.createTempFile(TMP_FILE_PREFIX, null);
+            tmpOutStream = new FileOutputStream(tmpTargetFile);
+            byte[] buffer = new byte[1024];
+            int readBytesTotal = 0;
+            InputStream httpInputStream = asyncListener.getInputStream();
```

Comment:
Shouldn't you close this after reading from it? Use try-with-resources if you can.

Reply:
OK